include/boost/beast/core/detail/impl/temporary_buffer.ipp

```diff
@@ -0,0 +1,73 @@
+//
+// Copyright (c) 2019 Damian Jarek(damian.jarek93@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef BOOST_BEAST_DETAIL_IMPL_TEMPORARY_BUFFER_IPP
+#define BOOST_BEAST_DETAIL_IMPL_TEMPORARY_BUFFER_IPP
+
+#include <boost/beast/core/detail/temporary_buffer.hpp>
+#include <boost/beast/core/detail/clamp.hpp>
+#include <boost/core/exchange.hpp>
+
+#include <memory>
+#include <cstring>
+
+namespace boost {
+namespace beast {
+namespace detail {
+
+
+void
+temporary_buffer::append(string_view sv)
+{
+    reserve_additional(sv.size());
+    unsafe_append(sv);
+}
+
+void
+temporary_buffer::append(string_view sv1, string_view sv2)
+{
+    reserve_additional(sv1.size() + sv2.size());
+    unsafe_append(sv1);
+    unsafe_append(sv2);
+}
+
+void
+temporary_buffer::unsafe_append(string_view sv)
+{
+    auto n = sv.size();
+    std::memcpy(&data_[size_], sv.data(), n);
+    size_ += n;
+}
+
+void
+temporary_buffer::reserve_additional(std::size_t sv_size)
+{
+    if (capacity_ - size_ >= sv_size)
+    {
+        return;
+    }
+
+    auto constexpr limit = (std::numeric_limits<std::size_t>::max)();
+    if (beast::detail::sum_exceeds(capacity_, capacity_, limit) ||
```

Comment:
This works but it would be better if it clamped the new capacity to `limit` instead of just throwing an exception when `capacity*2` is too large. I admit I tried to rearrange the code to achieve this but the results were all messy.

Reply:
I reworked the logic a bit to simplify the overflow checks. I also noticed that `basic_fields` currently assumes `std::size_t` is larger than `std::uint16_t`, so I've removed the overflow check in `temporary_buffer`, since it can never overflow, because the input we filter out fits in `std::uint16_t` and the output never exceeds `2*std::numeric_limits<std::uint16_t>::max() + small constant`.